### ISSUE TITLE
iOS Undebuggable wrong binary

### ISSUE THREAD

**Issue by chq-matteo** [commjoen/uncrackable_app#10]
# Contains spoilers for the challenge

## Error in current deployed .ipa

It looks to me that there is an error in the .ipa file in OWASP/owasp-mstg for the iOS Level 2 crackme.
https://github.com/OWASP/owasp-mstg/commit/1156664f02dbd07120d3909749b97ca4c7020108

There are a couple of issues that make me think that the there is an error.

1. The 64bit version and 32bit version have the same ciphertext (the so called solution base64 string 9Ing3PQpLWF3ILQRL3W/5yJgo+VQnJzkFjPnGQa/wxw=)
2. The solution string is different from the one in the source code found in this repo and also different from the one used in the only writeup (which is the same one we can find in this repo)

The binary in the owasp repo has the same checksum as the one that can be found in the ioslvl2update branch

```
e228d489bd2a7e4e35f0b082ce2bf797  UnCrackable_Level_2.ipa
e228d489bd2a7e4e35f0b082ce2bf797  UnDebuggable.ipa
```

However the ciphertext is different from the one found in the source code in that branch
```
https://github.com/commjoen/uncrackable_app/blob/ioslvl2update/iOS/Level2/UnDebuggable/ViewController.h
```

---

Based on these information, my guess is that one or more of these things happened:
- there has been a copy paste error somewhere
- the string belongs to an older version
- there were differences in the builds (e.g. signature computed on debug build then binary published in release mode)
- the wrong version was uploaded

---

## 64 bit support

On 64 bit devices the current ipa cannot work because the macho header parsing logic is wrong. It looks like the 64 bit macho header uses different constants because segments and sections have different fields and sizes compared to 32bit (LC_SEGMENT is equal to 1, but that means a 32 bit segment)

https://github.com/commjoen/uncrackable_app/blob/ioslvl2update/iOS/Level2/UnDebuggable/ViewController.m#L62

![image](https://user-images.githubusercontent.com/6654800/72474162-80bc8400-37e8-11ea-9a59-f382466e7421.png)

---

This is probably technical debt from a time when there weren't any 64 bit iDevice (not too long ago actually).

Since cross platform-ness of obfuscation/anti-debugging checks is not easily achieved, it would be reasonable to drop the 32 bit version since it is legacy for Apple anyways.

---

Thank you for maintaining this cool project

**Comment by chq-matteo** [comment]
What was proposed here https://github.com/OWASP/MSTG-Hacking-Playground/issues/4 would really help in this case, since at the moment building the level 2 app has a lot of possible pitfalls

**Comment by chq-matteo** [comment]
I forgot to say that since the decryption uses pkcs7 padding it should also be impossible to solve this challenge with dynamic analysis, since the decryption should error out

https://crypto.stackexchange.com/questions/9043/what-is-the-difference-between-pkcs5-padding-and-pkcs7-padding

**Comment by commjoen** [comment]
Hi @chq-matteo , thank you for filing a bug. Let me get back to you about this. Allthough that might take some time due to personal reasons.

**Comment by commjoen** [comment]
@sushi2k would you be able to help with this?

**Comment by sushi2k** [comment]
Hi @chq-matteo. Thanks for your detailed report, that is highly appreciated. Unfortunately as of now it's not possible to maintain the iOS Crackmes at the moment. We hope we can prioritise this after some other tasks of the MSTG/MASVS projects have been finished. Thanks again!

**Comment by chq-matteo** [comment]
sure, no problem

The crackme concept was cool, but I can see why it is impossible to mantain

I'll leave this open in case anyone else stumbles on this issue

**Comment by sushi2k** [comment]
Definitely. Let's leave it open, your description is very helpful. Thanks again.
